ddtrace/tracer/textmap.go

```diff
@@ -264,46 +265,71 @@ func (p *chainedPropagator) Inject(spanCtx ddtrace.SpanContext, carrier interfac
+// be relayed in the original trace context with a SpanLink.
 func (p *chainedPropagator) Extract(carrier interface{}) (ddtrace.SpanContext, error) {
-	var ctx ddtrace.SpanContext
+	var ctx ddtrace.SpanContext // First valid trace context will be stored here
+	var links []ddtrace.SpanLink
 	for _, v := range p.extractors {
-		if ctx != nil {
-			// A local trace context has already been extracted.
-			pw3c, isW3C := v.(*propagatorW3c)
-			if !isW3C {
-				continue // Ignore other propagators.
+		extractedCtx, err := v.Extract(carrier)
+		// Handling extraction errors. If it is the first extraction, return error. Else, ignore
+		if err != nil {
+			if ctx == nil && err != ErrSpanContextNotFound {
+				return nil, err
+			} else {
+				continue
 			}
-			w3cCtx, err := pw3c.Extract(carrier)
-			if err == nil && w3cCtx.(*spanContext).TraceID128() == ctx.(*spanContext).TraceID128() {
-				pw3c.propagateTracestate(ctx.(*spanContext), w3cCtx.(*spanContext))
-				if w3cCtx.SpanID() != ctx.SpanID() {
-					var ddCtx *spanContext
-					if ddp := getDatadogPropagator(p); ddp != nil {
-						if ddSpanCtx, err := ddp.Extract(carrier); err == nil {
-							ddCtx, _ = ddSpanCtx.(*spanContext)
+		}
+		if ctx != nil { // A local trace context has already been extracted
+			propagatorType, err := getPropagatorType(v)
+			if err != nil { // Propagator is not one of the known, supported types. Ignore this extraction
+				log.Debug("Propagator type is unknown. Skipping extraction")
+				continue
+			}
+			if extractedCtx.(ddtrace.SpanContextW3C).TraceID128() == ctx.(ddtrace.SpanContextW3C).TraceID128() {
+				if propagatorType == "tracecontext" {
+					v.(*propagatorW3c).propagateTracestate(ctx.(*spanContext), extractedCtx.(*spanContext)) //Sending tracestate to the ctx to be returned
+					if extractedCtx.SpanID() != ctx.SpanID() {
+						var ddCtx *spanContext
+						if ddp := getDatadogPropagator(p); ddp != nil {
+							if ddSpanCtx, err := ddp.Extract(carrier); err == nil {
+								ddCtx, _ = ddSpanCtx.(*spanContext)
+							}
 						}
+						overrideDatadogParentID(ctx.(*spanContext), extractedCtx.(*spanContext), ddCtx)
+					}
+				}
+			} else {
+				var flags uint32
+				if tracer := extractedCtx.(*spanContext).trace; tracer != nil {
+					if flags = uint32(*tracer.priority); flags > 0 { // Set the flags based on the sampling priority
+						flags = 1
+					} else {
+						flags = 0
 					}
-					overrideDatadogParentID(ctx.(*spanContext), w3cCtx.(*spanContext), ddCtx)
 				}
+				link := ddtrace.SpanLink{TraceID: extractedCtx.(*spanContext).TraceID(), TraceIDHigh: extractedCtx.(*spanContext).TraceIDUpper(), SpanID: extractedCtx.(*spanContext).SpanID(), Attributes: map[string]string{"reason": "terminated_context", "context_headers": propagatorType}, Flags: flags}
+				if propagatorType == "tracecontext" {
+					link.Tracestate = extractedCtx.(*spanContext).trace.propagatingTag(tracestateHeader)
+				}
+				links = append(links, link)
```

Comment:
As far as I can tell, this code block runs only if the following conditions have been met: 

1. First extraction has already occurred
2. We're on second+ extraction
3. And this propagator type is not w3c

Are you sure that's where you want spanlinks and line 316 to live? Shouldn't this logic happen regardless of # 3 ?

This might need to move outside of the most recent `else`.

Reply:
The way the code is written right now, this block will run as long as traceids of the extracted and local context are different. The suggestion you made above would mean that the span link logic would only occur if propagator type is not w3c. Thus, maybe we should keep the above check for traceids and propagator types separate?